### PR TITLE
restaking: added unbonding period to restaking

### PR DESCRIPTION
**Pull request by dhruvja**
Added 7 days of unbonding period to withdrawals. User has to request for withdrawal where their receipt token is transferred to escrow until the unbonding period is over. But during that period, user has an option to cancel the withdrawal request and get back their receipt token. Once the unbonding period is over, the `withdraw` method can be called to perform the withdrawal.

### CODE REVIEW

**Review by mina86 — APPROVED** [review]

File: solana/restaking/programs/restaking/src/constants.rs [hunk 1]
```diff
@@ -2,9 +2,18 @@ pub const STAKING_PARAMS_SEED: &[u8] = b"staking_params";
...
 pub const TOKEN_URI: &str =
     "https://arweave.net/QbxPlvN1nHFG0AVXfGNdlXUk-LEkrQxFffI3fOUDciA";
+
+#[cfg(not(feature = "mocks"))]
+/// 7 days in seconds (60 * 60 * 24 * 7)
+pub const UNBONDING_PERIOD_IN_SEC: u64 = 604800;
+
+#[cfg(feature = "mocks")]
+/// 1 second for tests
+pub const UNBONDING_PERIOD_IN_SEC: u64 = 1;
```
[the remark this answers is not in the document]
Comment (dhruvja) [reply solana/restaking/programs/restaking/src/constants.rs:19]: oh gotcha, sounds good.

File: solana/restaking/programs/restaking/src/lib.rs [hunk 2]
```diff
@@ -776,6 +1062,16 @@ pub enum Service {
     GuestChain { validator: Pubkey },
 }
 
+#[derive(AnchorDeserialize, AnchorSerialize, Clone, Debug, Copy)]
+pub struct WithdrawalRequestParams {
+    /// Timestamp when withdrawal was requested
+    timestamp_in_sec: u64,
```
[the remark this answers is not in the document]
Comment (dhruvja) [reply solana/restaking/programs/restaking/src/lib.rs:1068]: well i guess wouldnt hurt to add it.

**Comment by dhruvja** [comment]
Well looks like the IDL breaks if we use `NonZero`. Lets go with u64 for now and we can circle back later when this is merged https://github.com/coral-xyz/anchor/pull/2824

File: solana/restaking/programs/restaking/src/lib.rs [hunk 3]
```diff
@@ -1135,4 +1135,6 @@ pub enum ErrorCodes {
          withdrawal"
     )]
     InvalidWithdrawer,
+    #[msg("Expected a non zero timestamp. Found zero timestamp")]
+    UnexpectedZeroTimestamp,
```
Comment (mina86) [inline solana/restaking/programs/restaking/src/lib.rs:1139]:
```suggestion
```

**Pull request merged**